File: buddy-strings/solution859.go
// https://leetcode.com/problems/buddy-strings/

package main

import "fmt"

// T: O(n), S: O(1)
func buddyStrings(s string, goal string) bool {
	if len(s) != len(goal) {
		return false
	}
	// Edge case
	if s == goal {
		counts := [26]byte{}
		for _, ch := range s {
			counts[ch-'a']++
			if counts[ch-'a'] == 2 {
				return true
			}
		}
		return false
	}
	n, diffCount := len(s), 0
	sCounts, gCounts := [26]int{}, [26]int{}
	for i := 0; i < n; i++ {
		sCounts[s[i]-'a']++
		gCounts[goal[i]-'a']++
		if s[i] != goal[i] {
			diffCount++
			if diffCount > 2 {
				return false
			}
		}
	}
	return diffCount == 2 && sCounts == gCounts
}

func main() {
	fmt.Println(buddyStrings("ab", "ba")) // true
	fmt.Println(buddyStrings("ab", "ab")) // false
	fmt.Println(buddyStrings("aa", "aa")) // true
}

File: buddy-strings/Solution859.java
// https://leetcode.com/problems/buddy-strings/

import java.util.Arrays;

class Solution859 {

    // T: O(n), S: O(1)
    public boolean buddyStrings(String s, String goal) {
        if (s.length() != goal.length()) {
            return false;
        }
        // Edge case
        if (s.equals(goal)) {
            var counts = new int[26];
            for (char ch : s.toCharArray()) {
                counts[ch - 'a']++;
                if (counts[ch - 'a'] == 2) {
                    return true;
                }
            }
            return false;
        }
        int n = s.length(), diffCount = 0;
        int[] sCounts = new int[26], gCounts = new int[26];
        for (int i = 0; i < n; i++) {
            sCounts[s.charAt(i) - 'a']++;
            gCounts[goal.charAt(i) - 'a']++;
            if (s.charAt(i) != goal.charAt(i)) {
                diffCount++;
                if (diffCount > 2) {
                    return false;
                }
            }
        }
        return diffCount == 2 && Arrays.equals(sCounts, gCounts);
    }

    public static void main(String[] args) {
        var sol = new Solution859();
        System.out.println(sol.buddyStrings("ab", "ba")); // true
        System.out.println(sol.buddyStrings("ab", "ab")); // false
        System.out.println(sol.buddyStrings("aa", "aa")); // true
    }

}
